src/socket/mod.rs

```diff
@@ -0,0 +1,311 @@
+//! Talking to the `lorri` daemon / unix sockets.
+
+pub mod communicate;
+pub mod path;
+
+use std::convert::TryFrom;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::os::unix::net::UnixStream;
+use std::time::{Duration, Instant};
+
+/// Wrapper around a socket that can send and receive structured messages.
+///
+/// `timeout` arguments set the socket timeout before reading/writing.
+pub struct ReadWriter<'a, R, W> {
+    // where R: serde::Deserialize {
+    socket: &'a UnixStream,
+    phantom_r: PhantomData<R>,
+    phantom_w: PhantomData<W>,
+}
+
+/// Milliseconds accepted by a `Timeout`.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Millis(u16);
```

Comment:
Hmm why make our own when Duration can work okay too?

Reply:
Originally it was like that, but then I found out that the `poll` syscall can only accept timeouts of up to 16 bit ms length. So instead of making every single read/write return a `Result`, rather make it impossible to construct a timeout that is out of bounds.